LIBFCREPO-1708. Created amistuck.sh script to detect stuck queues.

peichman-umd opened the pull request:
This script is intended for use as a liveness probe of the ActiveMQ pod.

https://umd-dit.atlassian.net/browse/LIBFCREPO-1708

peichman-umd committed:
LIBFCREPO-1708. Created amistuck.sh script to detect stuck queues.

This script is intended for use as a liveness probe of the ActiveMQ pod.

https://umd-dit.atlassian.net/browse/LIBFCREPO-1708

dbowling commented on amistuck.sh:
```diff
@@ -0,0 +1,53 @@
+# time the size was recorded as changing. If that value is greater than the
+# threshold, the script emits a status message to STDOUT and exits with a
+# nonzero status.
+
```
Universally a good thing to add to bash scripts:

```suggestion
set -euo pipefail
```

peichman-umd committed:
Fixed syntax error

dbowling commented on amistuck.sh:
```diff
@@ -0,0 +1,53 @@
+THRESHOLD=${2:-300}
+
+URL="http://localhost:8161/api/jolokia/read/org.apache.activemq:brokerName=localhost,destinationName=$QUEUE,destinationType=Queue,type=Broker"
+SIZE_FILE="/tmp/activemq/$QUEUE"
```
This will help in debugging if there's a read only root fs error.

```suggestion
SIZE_FILE="/tmp/activemq/$QUEUE"

check_writable() {
    local file="${1:-}"
    if [[ -e "${file}" ]]; then
        # File exists — check direct write permission
        if [[ ! -w "${file}" ]]; then
            echo "File '${file}' exists but is not writable." >&2
            exit 1
        fi
    else
        # File does not exist — check parent directory
        local dir
        dir="$(dirname "${file}")"
        if [[ ! -d "${dir}" || ! -w "${dir}" ]]; then
            echo "File '${file}' does not exist and directory '${dir}' is not writable." >&2
            exit 1
        fi
    fi
}
check_writable "${SIZE_FILE}"
```

peichman-umd committed:
Updated documentation comment.

dbowling commented on amistuck.sh:
```diff
@@ -0,0 +1,53 @@
+      fi
+    fi
+  fi
+fi
```
You certainly don't have to accept this rewrite, I would have left it untouched except that this script is probably going to be an example for others pointed out in the docs, so a few practices I prefer have been suggested:

- Logging
	- Logging in `key=value` pairs make for easy filtering, especially in Splunk.
		- This could be mixed, e.g. `severity=error queue="myqueue" queue has been stuck for greater than threshold...` is totally valid, though I often make all (or nearly all) in the `key=value` format when I can because if it is worth logging, it probably will be worth being able to filter/chart.
	- Clearly mark each message severity as an explicit `error`, `warning`, or `info` 
  		- `severity=error` format is just for Splunk benefits
  		- messages like `ERROR: [message]...` are fine too 
	- I presume you would want the pod log to indicate the warning state prior to the threshold being met, so I added that 
- Refactoring to remove nested ifs can help prevent bugs


Also, I tend to add some logging even for known good states for visibility. It will be interesting to see before we merge this in is if echo to the stdout and stderr go anywhere at all that is visible to us when the check passes (and exits `0`), as I think the kubelet may not persist them to a log unless verbosity is set to a high level. Either way, it can be nice if needed, but we probably don't want to be echoing those info level messages back to the pod, so I have it going to stdout exclusively. Please let me know how that goes during your tests so I can make the Kubernetes docs we keep reflect that information.

```suggestion
# Establish baseline values
last_size=$(cat "$SIZE_FILE" 2>/dev/null)
current_size=$(curl -s -H 'Origin: http://localhost' "$URL" | grep -o -e '"QueueSize":[0-9]*' | cut -d: -f2)
now=$(date +'%s')
last_change=$(stat -c'%Y' "$SIZE_FILE")
elapsed=$((now - last_change))

# exit early when the queue is empty
if [ "$current_size" -eq 0 ]; then
    echo "severity=info queue=\"${QUEUE}\" msg=\"queue is empty\" elapsed=\"${elapsed}\" last_size=\"${last_size:-n/a}\" current_size=\"${current_size}\""
    exit 0
fi

# update state and exit early when the queue size is changing
if [ "$current_size" != "$last_size" ]; then
    echo "severity=info queue=\"${QUEUE}\" msg=\"queue is changing, an updated size will be recorded\" elapsed=\"${elapsed}\" last_size=\"${last_size:-n/a}\" current_size=\"${current_size}\""
    # only when it is changed do we update the size file (to preserve the timestamp of the change)
    echo "$current_size" > "$SIZE_FILE"
    exit 0
fi

# size remaining at a non-zero value longer than $THRESHOLD is NOT okay
if (( elapsed > THRESHOLD )); then
    error="severity=error queue=\"${QUEUE}\" msg=\"queue has been stuck for greater than threshold ${THRESHOLD} seconds\" elapsed=\"${elapsed}\" last_size=\"${last_size:-n/a}\" current_size=\"${current_size}\""
    # send the error message to both the pod log (file descriptor 1 of pid 1)...
    echo "${error}" > /proc/1/fd/1
    # ...and the liveness probe failure message (STDERR)
    echo "${error}" >&2
    exit 1
fi

# if we get here, the queue is not changing but it hasn't exceeded the threshold yet
warning="severity=warning queue=\"${QUEUE}\" msg=\"queue is not changing, but threshold of ${THRESHOLD} seconds has not yet been exceeded\" elapsed=\"${elapsed}\" last_size=\"${last_size:-n/a}\" current_size=\"${current_size}\""
# send the warning message to both the pod log (file descriptor 1 of pid 1) and the liveness probe log (stderr)...
echo "${warning}" > /proc/1/fd/1
echo "${warning}" >&2
```

peichman-umd replied:
@dbowling I like the clarity of this rewrite. I am wondering if there is a reason to prefer the `[ "$current_size" -eq 0 ]` and `[ "$current_size" != "$last_size" ]` syntax that you used in lines 30 and 36 of this suggestions, as opposed to Bash numeric context style: `(( current_size == 0 ))` and `(( current_size != last_size ))`?

dbowling replied:
@peichman-umd 

I can't think of one, that was an incidental change that was introduced since I was writing some lines from scratch, and likely a byproduct of some of my IDE's preferences on Bash (I had to fight Shellcheck in a couple places to not needlessly make more changes based on `[[` and `${var}` everywhere.) Frankly, I also sometimes forget that Bash has numeric comparisons. 🤦

I think what you had on those is better, so here's an updated diff setting that back (though GitHub is doing worse formatting for it):

```suggestion
# Establish baseline values
last_size=$(cat "$SIZE_FILE" 2>/dev/null)
current_size=$(curl -s -H 'Origin: http://localhost/' "$URL" | grep -o -e '"QueueSize":[0-9]*' | cut -d: -f2)
now=$(date +'%s')
last_change=$(stat -c'%Y' "$SIZE_FILE")
elapsed=$((now - last_change))
# exit early when the queue is empty
if (( current_size == 0 )); then
    echo "severity=info queue=\"${QUEUE}\" msg=\"queue is empty\" elapsed=\"${elapsed}\" last_size=\"${last_size:-n/a}\" current_size=\"${current_size}\""
    exit 0
fi
# update state and exit early when the queue size is changing
if (( current_size != last_size )); then
    echo "severity=info queue=\"${QUEUE}\" msg=\"queue is changing, an updated size will be recorded\" elapsed=\"${elapsed}\" last_size=\"${last_size:-n/a}\" current_size=\"${current_size}\""
    # only when it is changed do we update the size file (to preserve the timestamp of the change)
    echo "$current_size" > "$SIZE_FILE"
    exit 0
fi
# size remaining at a non-zero value longer than $THRESHOLD is NOT okay
if (( elapsed > THRESHOLD )); then
    error="severity=error queue=\"${QUEUE}\" msg=\"queue has been stuck for greater than threshold ${THRESHOLD} seconds\" elapsed=\"${elapsed}\" last_size=\"${last_size:-n/a}\" current_size=\"${current_size}\""
    # send the error message to both the pod log (file descriptor 1 of pid 1)...
    echo "${error}" > /proc/1/fd/1
    # ...and the liveness probe failure message (STDERR)
    echo "${error}" >&2
    exit 1
fi
# if we get here, the queue is not changing but it hasn't exceeded the threshold yet
warning="severity=warning queue=\"${QUEUE}\" msg=\"queue is not changing, but threshold of ${THRESHOLD} seconds has not yet been exceeded\" elapsed=\"${elapsed}\" last_size=\"${last_size:-n/a}\" current_size=\"${current_size}\""
# send the warning message to both the pod log (file descriptor 1 of pid 1) and the liveness probe log (stderr)...
echo "${warning}" > /proc/1/fd/1
echo "${warning}" >&2
```

dbowling reviewed (changes requested):
@peichman-umd So excited to see this!

- the full PR is very mergable without accepting my changes.
- `set -eou pipefail` is a great setting to put in all bash scripts (though often forgotten.)
- `check_writable` is probably smart to get good error messages, especially in the context of a container with read only root FS.
- Please consider the giant diff suggestion _very_ optional. Read it and do what you wish with the info.
- Please test what you accept, I wrote the suggestions without applying and running.

peichman-umd committed:
Add some script safety with `set -eou pipefail`.

Co-authored-by: Dan Bowling <dbowling@users.noreply.github.com>

peichman-umd committed:
Added `check_writable` function to ensure the size file is writeable.

Helps to catch problems related to a read-only root file system.

Co-authored-by: Dan Bowling <dbowling@users.noreply.github.com>

peichman-umd committed:
LIBFCREPO-1708. Updates for easier readability.

- removed nested conditionals and rely on a "check condition and exit early" strategy
- created a `log` function to encapsulate either sending to STDOUT or to the pod log and STDERR based on the log severity
- check for at least 1 argument and exit with usage information if not present

https://umd-dit.atlassian.net/browse/LIBFCREPO-1708

peichman-umd commented:
@dbowling I've made updates incorporating a large part of your rewrite, plus an additional `log` function and checking for the minimum number of arguments.

dbowling reviewed (approved):
LGTM. Good change.

peichman-umd committed:
LIBFCREPO-1708. Adjusted the order of checks.

- check for and record a change before checking for a current value of 0, to have a record of when the queue went to 0

https://umd-dit.atlassian.net/browse/LIBFCREPO-1708
